x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -164,6 +165,18 @@ ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, IndicesR
         final ResolvedIndices split;
         if (indicesRequest instanceof IndicesRequest.SingleIndexNoWildcards single && single.allowsRemoteIndices()) {
             split = remoteClusterResolver.splitLocalAndRemoteIndexNames(indicesRequest.indices());
+            // all indices can come back empty when the remote index expression included a cluster alias with a wildcard
+            // and no remote clusters are configured that match it
+            if (split.getLocal().isEmpty() && split.getRemote().isEmpty()) {
+                for (String indexExpression : indices) {
+                    String[] clusterAndIndex = indexExpression.split(":", 2);
+                    if (clusterAndIndex.length == 2) {
+                        if (clusterAndIndex[0].contains("*")) {
+                            throw new NoSuchRemoteClusterException(clusterAndIndex[0]);
+                        }
+                    }
+                }
+            }
```

Comment:
I see that `NoSuchRemoteClusterException` is used in code for concrete remote cluster names (not cluster name wildcards).

Also, in many other places we just use `IndexNotFoundException`s, so I would prefer this simpler version:
```
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
index 42a1d89a9aa..99359089e37 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
@@ -167,15 +167,8 @@ class IndicesAndAliasesResolver {
             split = remoteClusterResolver.splitLocalAndRemoteIndexNames(indicesRequest.indices());
             // all indices can come back empty when the remote index expression included a cluster alias with a wildcard
             // and no remote clusters are configured that match it
-            if (split.getLocal().isEmpty() && split.getRemote().isEmpty()) {
-                for (String indexExpression : indices) {
-                    String[] clusterAndIndex = indexExpression.split(":", 2);
-                    if (clusterAndIndex.length == 2) {
-                        if (clusterAndIndex[0].contains("*")) {
-                            throw new NoSuchRemoteClusterException(clusterAndIndex[0]);
-                        }
-                    }
-                }
+            if (split.isEmpty()) {
+                throw new IndexNotFoundException(Arrays.toString(indicesRequest.indices()));
             }
         } else {
             split = new ResolvedIndices(Arrays.asList(indicesRequest.indices()), List.of());
``` 

But, more importantly, what is the behavior when Security is disabled, when `_scripts/painless/_execute` goes for a remote cluster name wildcard that doesn't cover any configured cluster names? I think this is the error we should also throw here.

Reply:
> But, more importantly, what is the behavior when Security is disabled, when _scripts/painless/_execute goes for a remote cluster name wildcard that doesn't cover any configured cluster names?

When security is disabled and a wildcard is used that does not match a remote cluster, `NoSuchRemoteClusterException` is thrown:

```
 curl -s -D /dev/stderr -w "\n" -H "Content-Type: application/json" -X POST localhost:9200/_scripts/painless/_execute --data '{"script":{"source":"doc[\"field\"].value.length() <= params.max_length","params":{"max_length":4}},"context":"filter","context_setup":{"index":"w*w:test","document":{"field":"four"}}}'

HTTP/1.1 404 Not Found
X-elastic-product: Elasticsearch
content-type: application/json
content-length: 207

{"error":{"root_cause":[{"type":"no_such_remote_cluster_exception","reason":"no such remote cluster: [w*w]"}],"type":"no_such_remote_cluster_exception","reason":"no such remote cluster: [w*w]"},"status":404}
```

(I tested the above on both `main` and this branch to make sure I didn't change that behavior as well).

So with this PR, both routes (security enabled or disabled) return the same error message.

In the latest push, I have updated the painless API docs to indicate that wildcards are not allowed in either the cluster alias or the index portion of the index expression and list what error to expect. See [this commit](https://github.com/elastic/elasticsearch/pull/109185/commits/d0d913bf8035a0c72ec4ac8b3730e9146d8344b).